src/EarlyFree.cpp

```diff
@@ -44,6 +47,9 @@ class FindLastUse : public IRVisitor {
         if (call->name == func) {
             last_use = containing_stmt;
         }
+        if (call->name == "halide_device_malloc" && expr_uses_var(call, func + ".buffer")) {
```

Comment:
I don't think so, that shouldn't matter. Basically, if the buffer ever had any device_malloc on it, regardless of device interface, we should call device_free. It's really doing the same thing as destructor would, just earlier.